### DISCUSSION TITLE
MultiValueDictionary

### DISCUSSION THREAD

**Discussion by roederja** [dotnet/dotNext#163] [Ideas]
Is there a chance to include the MultiValueDictionary class from the old Microsoft.Experimental.Collections package in dotNext: https://devblogs.microsoft.com/dotnet/multidictionary-becomes-multivaluedictionary/

Or is there a reason to keep it out? I personally found it very useful in production code and safes you from writing boilerplate code for Dictionaries where the value is a list or set.

**Comment by sakno** [comment]
It looks like a corner case for dictionary. MultiValueDictionary can be represented by `Dictionary<TKey, List<TValue>>`. It was implemented by Microsoft.Experimental.Collections in the same way. So I see no reason to introduce a new class, but it is possible to convert some boilerplate code to something reusable using extension methods, e.g.
```csharp
public static IReadOnlyCollection<TValue> TryGetValues<TKey, TValue, TCollection>(this IDictionary<TKey, TCollection> dictionary)
  where TKey : notnull
  where TCollection : class, IReadOnlyCollection<TValue>;
```

Is it reasonable for you? If yes, you can provide a set methods you want to see in the library as a proposal.

**Reply by roederja** [reply]
My gut reaction is that this would be awkward to use and would pollute the Dictionary namespace.

One limitation of the extension method idea is that you have to use a concrete collection type in the declaration or you'd always have to explicitly specify the collection type on the Add method, so the extension method knows what type of collection to create - clunky and dangerous because you could end up with some elements being lists and some being sets.
Furthermore you'd have two Add methods: the original Add method that accepts a collection type as the value (List<TValue>) and the extentions method which accepts the TValue (from your example above). I think that would be quite confusing.

**Reply by sakno** [reply]
> have to use a concrete collection type in the declaration

Right, because the dictionary need to know how to store multiple values - list or set.

> to explicitly specify the collcetion type on the Add method

You don't need. Compiler can do this work for you. For instance:
```csharp
static class MutliValueDictionary
{
    public static TCollection Add<TKey, TValue, TCollection>(this IDictionary<TKey, TCollection> dictionary, TKey key, TValue value)
        where TKey : notnull
        where TCollection : class, ICollection<TValue>, new()
    {
        if (!dictionary.TryGetValue(key, out TCollection collection))
            dictionary.Add(key, collection = new());

        collection.Add(value);
        return collection;
    }
}

Dictionary<string, List<int>> dictionary = new();
dictionary.Add(string.Empty, 42); // resolves as extension method
```

> dangerous because you could end up with some elements being lists and some being sets

Not sure if it's possible with my example above.

> Furthermore you'd have two Add methods

That's true. However, it makes sense if you want to add multiple values by single call.

**Reply by roederja** [reply]
> You don't need. Compiler can do this work for you. For instance:

Yes, but in this case TCollection needs to be a concrete type - not an interface like IReadOnlyCollection

> However, it makes sense if you want to add multiple values by single call.

But in this case you'd want them to be added to what's already there. The Dictionary Add method fails when the key already exists.

I think these extension methods would have too many quirks to be added to a library like dotNext. Microsoft might have had a reason not to include their MultiValueDictionary into the BCL in the end - I just don't know what they were or if they just forgot about it. I thought it might be a good addtion to dotNext, but if you disagree I can accept that.

**Reply by sakno** [reply]
I just don't want to copy-paste from MS sources (which is probably forbidden by the license). This is why I'm trying to figure out real use cases. One of them - HTTP headers. The same header may have multiple values. But ASP.NET Core provides its own collection type for the headers.

**Reply by roederja** [reply]
Oh I see - I assumed the Microsft code is open source? https://github.com/dotnet/corefxlab/blob/e0c8705741ad2b66a1bbc70ff5bf5e02886eca1e/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/MultiValueDictionary.cs#L12

looks like it's under the MIT license.

**Reply by sakno** [reply]
MIT license means that I need to mention the original work in case of derivative work.

**Reply by roederja** [reply]
Is that a problem?

**Reply by sakno** [reply]
Nope, but I prefer to implement things without looking at third-party code to avoid any potential issues with the license. This requires some effort, this is why I'm thinking about potential use cases.

**Reply by roederja** [reply]
I see. That would be quite a lot of effort to get to something of the same standard as the existing code. I think it was an intern project and probably at least a month of work went into this.

**Reply by sakno** [reply]
Additional considerations: almost every collection type in BCL has companion type. For instance, `Dictionary<K, V>` has `ImmutableDictionary<K, V>`, `FrozenDictionary<K, V>` and `ConcurrentDictionary<K, V>`. Should we have the same for `MultiValueDictionary`?

**Reply by roederja** [reply]
There is already the Lookup type which is essentially a read only version of a MultiValueDictionary.

**Reply by sakno** [reply]
Read only is not the same as immutable. Typically, immutable types provide modification methods with _copy-on-write_ semantics.

**Reply by roederja** [reply]
I think there is much less of a case for an immutable multi value dictionary. The boiler plate for constructing that out of the existing immutable types is fairly managable.

**Reply by sakno** [reply]
> The boiler plate for constructing that out of the existing immutable types is fairly managable.

But not for mutable? Also there is thread-safe companion of `Dictionary<K, V>`.

**Reply by roederja** [reply]
I would say out of the map types dictionary is by far the most commonly used - so that should be the main candidate for a specialised multi value version.

**Reply by sakno** [reply]
`MultiValueConcurrentDictionary` looks more promising. It can be used as simple multi-value dictionary as well. However, `ConcurrentDictionary` from BCL cannot be used to thread-safe multi-value dictionary while classic `Dictionary` can be used to emulate thread-unsafe multi-value dictionary.

**Reply by roederja** [reply]
I think it will be very difficult to create a lock free multi value dictionary - if it’s possible at all.

**Reply by sakno** [reply]
`ConcurrentDictionary` is not lock-free. In case of reads, yes, it is. But in case of writes - no. It just effectively reduces the chance of lock contention, no more. You can take a look at its source code and you'll see the presence of `Monitor.Enter`.

**Reply by sakno** [reply]
Also, I found some ambiguity with API surface. For instance,
```csharp
bool TryAdd(TKey key, TValue value);
```

What this method should do? Don't add a value to the list if `key` is already exist (just like regular `Dictionary`) or try to add the value to the list/set and return **false** if `value` is already exist.

**Reply by roederja** [reply]
The MultiValueDictionary from Microsoft doesn't have TryAdd - probably for that reason.

**Reply by sakno** [reply]
But what if dictionary created with an option to use `HashSet` as a storage for values?

**Reply by roederja** [reply]
It will always call Add on the underlying collection. In the list case you’d end up with duplicates - in the set case you won’t. Seems reasonable to me.

**Reply by sakno** [reply]
This is why `List<T>.Add` doesn't return anything while `HashSet<T>.Add` returns **bool**. It's good to know about duplicate value, however, in the context of MultiValueDictionary such a method produces ambiguity.

**Comment by ZzZombo** [comment]
Types with value equality semantics like `ValueTuple<>` or third-party ones can be used as keys for ordinal dictionary types.
